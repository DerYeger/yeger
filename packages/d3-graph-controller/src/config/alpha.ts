import type { NodeTypeToken } from '~/model/graph'
import type { GraphNode } from '~/model/node'

/**
 * Alpha values when label display changes.
 */
export interface LabelAlphas {
  /**
   * Alpha value when labels are turned off.
   */
  readonly hide: number
  /**
   * Alpha value when labels are turned on.
   */
  readonly show: number
}

/**
 * Context of a resize.
 */
export interface ResizeContext {
  /**
   * The old height.
   */
  readonly oldHeight: number
  /**
   * The old width.
   */
  readonly oldWidth: number
  /**
   * The new height.
   */
  readonly newHeight: number
  /**
   * The new width.
   */
  readonly newWidth: number
}

/**
 * Alpha value configuration for controlling simulation activity.
 */
export interface AlphaConfig<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
> {
  /**
   * Target alpha values for dragging.
   */
  readonly drag: {
    /**
     * Target alpha when a drag starts.
     * Should be larger than 0.
     */
    readonly start: number
    /**
     * Target alpha when a drag stops.
     * Should generally be 0.
     */
    readonly end: number
  }
  /**
   * Alpha values for filter changes.
   */
  readonly filter: {
    /**
     * Alpha value when the link filter changes.
     */
    readonly link: number
    /**
     * Alpha value when the node type filter changes.
     */
    readonly type: number
    /**
     * Alpha values when the inclusion of unlinked nodes changes.
     */
    readonly unlinked: {
      /**
       * Alpha value when unlinked nodes are included.
       */
      readonly include: number
      /**
       * Alpha value when unlinked nodes are excluded.
       */
      readonly exclude: number
    }
  }
  /**
   * Alpha values when node focus changes.
   */
  readonly focus: {
    /**
     * Alpha value when a node is focused.
     * @param node - The focused node.
     * @returns The alpha value.
     */
    readonly acquire: (node: Node) => number
    /**
     * Alpha value when a node is unfocused.
     * @param node - The unfocused node.
     * @returns The alpha value.
     */
    readonly release: (node: Node) => number
  }
  /**
   * Alpha value when the graph is initialized.
   */
  readonly initialize: number
  /**
   * Alpha values when label display changes.
   */
  readonly labels: {
    /**
     * Alpha values when link label display changes.
     */
    readonly links: LabelAlphas
    /**
     * Alpha values when node label display changes.
     */
    readonly nodes: LabelAlphas
  }
  /**
   * Alpha values when the graph is resized.
   */
  readonly resize: number | ((context: ResizeContext) => number)
}

/**
 * Create the default alpha configuration.
 */
export function createDefaultAlphaConfig<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
>(): AlphaConfig<T, Node> {
  return {
    drag: {
      end: 0,
      start: 0.1,
    },
    filter: {
      link: 1,
      type: 0.1,
      unlinked: {
        include: 0.1,
        exclude: 0.1,
      },
    },
    focus: {
      acquire: () => 0.1,
      release: () => 0.1,
    },
    initialize: 1,
    labels: {
      links: {
        hide: 0,
        show: 0,
      },
      nodes: {
        hide: 0,
        show: 0,
      },
    },
    resize: 0.5,
  }
}
