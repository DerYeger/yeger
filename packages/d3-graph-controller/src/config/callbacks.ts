import type { NodeTypeToken } from '~/model/graph'
import type { GraphNode } from '~/model/node'

/**
 * Callback configuration.
 */
export interface Callbacks<T extends NodeTypeToken, Node extends GraphNode<T>> {
  /**
   * Callback when a node is double-clicked or double-tapped.
   * @param node - The node.
   */
  readonly nodeClicked?: (node: Node) => void
}
