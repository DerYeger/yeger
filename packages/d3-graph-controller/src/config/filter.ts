import type { NodeTypeToken } from '~/model/graph'
import type { GraphLink } from '~/model/link'
import type { GraphNode } from '~/model/node'

/**
 * Link filter.
 * Receives a link and returns whether the link should be included or not.
 */
export type LinkFilter<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>,
> = (link: Link) => boolean
