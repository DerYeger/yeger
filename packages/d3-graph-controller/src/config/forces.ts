import type { NodeTypeToken } from '~/model/graph'
import type { GraphLink } from '~/model/link'
import type { GraphNode } from '~/model/node'

/**
 * Simulation force.
 */
export interface Force<Subject> {
  /**
   * Whether the force is enabled.
   */
  readonly enabled: boolean
  /**
   * The strength of the force.
   * Can be a static number or a function receiving the force's subject and returning a number.
   */
  readonly strength: number | ((subject: Subject) => number)
}

/**
 * Simulation force applied to nodes.
 */
export type NodeForce<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
> = Force<Node>

/**
 * Collision force applied to nodes.
 */
export interface CollisionForce<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
> extends NodeForce<T, Node> {
  /**
   * Multiplier of the node radius.
   * Tip: Large values can drastically reduce link intersection.
   */
  readonly radiusMultiplier: number
}

/**
 * Simulation force applied to links.
 */
export interface LinkForce<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>,
> extends Force<Link> {
  /**
   * Define the length of a link for the simulation.
   */
  readonly length: number | ((link: Link) => number)
}

/**
 * Simulation force configuration.
 */
export interface SimulationForceConfig<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>,
> {
  /**
   * Centering force applied to nodes.
   */
  readonly centering: false | NodeForce<T, Node>
  /**
   * Charge force applied to nodes.
   */
  readonly charge: false | NodeForce<T, Node>
  /**
   * Collision force applied to nodes.
   */
  readonly collision: false | CollisionForce<T, Node>
  /**
   * Link force applied to links.
   */
  readonly link: false | LinkForce<T, Node, Link>
}

/**
 * Create the default force configuration.
 */
export function createDefaultForceConfig<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>,
>(): SimulationForceConfig<T, Node, Link> {
  return {
    centering: {
      enabled: true,
      strength: 0.1,
    },
    charge: {
      enabled: true,
      strength: -1,
    },
    collision: {
      enabled: true,
      strength: 1,
      radiusMultiplier: 2,
    },
    link: {
      enabled: true,
      strength: 1,
      length: 128,
    },
  }
}
