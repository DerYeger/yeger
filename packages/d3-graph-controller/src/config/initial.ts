import type { LinkFilter } from '~/config/filter'
import type { NodeTypeToken } from '~/model/graph'
import type { GraphLink } from '~/model/link'
import type { GraphNode } from '~/model/node'

/**
 * Initial settings of a controller.
 */
export interface InitialGraphSettings<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>
> {
  /**
   * Whether nodes without incoming or outgoing links will be shown or not.
   */
  readonly includeUnlinked: boolean
  /**
   * Link filter that decides whether links should be included or not.
   */
  readonly linkFilter: LinkFilter<T, Node, Link>
  /**
   * Node types that should be included.
   * If undefined, all node types will be included.
   */
  readonly nodeTypeFilter?: T[] | undefined
  /**
   * Whether link labels are shown or not.
   */
  readonly showLinkLabels: boolean
  /**
   * Whether node labels are shown or not.
   */
  readonly showNodeLabels: boolean
}

/**
 * Create default initial settings.
 */
export function createDefaultInitialGraphSettings<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>
>(): InitialGraphSettings<T, Node, Link> {
  return {
    includeUnlinked: true,
    linkFilter: () => true,
    nodeTypeFilter: undefined,
    showLinkLabels: true,
    showNodeLabels: true,
  }
}
