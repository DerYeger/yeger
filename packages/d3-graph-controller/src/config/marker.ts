import type { GraphConfig } from './config'
import { getNodeRadius } from '../lib/utils'
import type { NodeTypeToken } from '../model/graph'
import type { GraphLink } from '../model/link'
import type { GraphNode } from '../model/node'

/**
 * Marker configuration.
 */
export interface MarkerConfig {
  /**
   * Size of the marker's box.
   */
  readonly size: number
  /**
   * Get padding of the marker for calculating link paths.
   * @param node - The node the marker is pointing at.
   * @param config - The current config.
   * @returns The padding of the marker.
   */
  readonly padding: <
    T extends NodeTypeToken,
    Node extends GraphNode<T>,
    Link extends GraphLink<T, Node>,
  >(
    node: Node,
    config: GraphConfig<T, Node, Link>,
  ) => number
  /**
   * The ref of the marker.
   */
  readonly ref: [number, number]
  /**
   * The path of the marker.
   */
  readonly path: [number, number][]
  /**
   * The ViewBox of the marker.
   */
  readonly viewBox: string
}

function defaultMarkerConfig(size: number): MarkerConfig {
  return {
    size,
    padding: <
      T extends NodeTypeToken,
      Node extends GraphNode<T>,
      Link extends GraphLink<T, Node>,
    >(
      node: Node,
      config: GraphConfig<T, Node, Link>,
    ) => getNodeRadius(config, node) + 2 * size,
    ref: [size / 2, size / 2],
    path: [
      [0, 0],
      [0, size],
      [size, size / 2],
    ] as [number, number][],
    viewBox: [0, 0, size, size].join(','),
  }
}

/**
 * Collection of built-in markers.
 */
export const Markers = {
  /**
   * Create an arrow marker configuration.
   * @param size - The size of the arrow
   */
  Arrow: (size: number): MarkerConfig => defaultMarkerConfig(size),
}
