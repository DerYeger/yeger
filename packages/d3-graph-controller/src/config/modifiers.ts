import type { Selection } from 'd3-selection'

import type { Drag, GraphSimulation, Zoom } from '../lib/types'
import type { NodeTypeToken } from '../model/graph'
import type { GraphLink } from '../model/link'
import type { GraphNode } from '../model/node'

/**
 * Modifier for the drag.
 */
export type DragModifier<T extends NodeTypeToken, Node extends GraphNode<T>> = (
  drag: Drag<T, Node>,
) => void

/**
 * Modifier for node circles.
 */
export type NodeModifier<T extends NodeTypeToken, Node extends GraphNode<T>> = (
  selection: Selection<SVGCircleElement, Node, SVGGElement, undefined>,
) => void

/**
 * Modifier for node labels.
 */
export type NodeLabelModifier<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
> = (selection: Selection<SVGTextElement, Node, SVGGElement, undefined>) => void

/**
 * Modifier for link paths.
 */
export type LinkModifier<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>,
> = (selection: Selection<SVGPathElement, Link, SVGGElement, undefined>) => void

/**
 * Modifier for link labels.
 */
export type LinkLabelModifier<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>,
> = (selection: Selection<SVGTextElement, Link, SVGGElement, undefined>) => void

/**
 * Modifier for the simulation.
 */
export type SimulationModifier<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>,
> = (simulation: GraphSimulation<T, Node, Link>) => void

/**
 * Modifier for the zoom.
 */
export type ZoomModifier = (zoom: Zoom) => void

/**
 * Low-level callbacks for modifying the underlying d3-selection.wd
 */
export interface Modifiers<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>,
> {
  /**
   * Modify the drag.
   * @param drag - The drag.
   */
  readonly drag?: DragModifier<T, Node>
  /**
   * Modify the node selection.
   * @param selection - The selection of nodes.
   */
  readonly node?: NodeModifier<T, Node>
  /**
   * Modify the node label selection.
   * @param selection - The selection of node labels.
   */
  readonly nodeLabel?: NodeLabelModifier<T, Node>
  /**
   * Modify the link selection.
   * @param selection - The selection of links.
   */
  readonly link?: LinkModifier<T, Node, Link>
  /**
   * Modify the link label selection.
   * @param selection - The selection of link labels.
   */
  readonly linkLabel?: LinkLabelModifier<T, Node, Link>
  /**
   * Modify the simulation.
   * @param simulation - The simulation.
   */
  readonly simulation?: SimulationModifier<T, Node, Link>
  /**
   * Modify the zoom.
   * @param zoom - The zoom.
   */
  readonly zoom?: ZoomModifier
}
