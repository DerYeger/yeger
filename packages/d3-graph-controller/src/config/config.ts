import merge from 'ts-deepmerge'

import { createDefaultAlphaConfig } from '~/config/alpha'
import type { Callbacks } from '~/config/callbacks'
import { createDefaultForceConfig } from '~/config/forces'
import type { InitialGraphSettings } from '~/config/initial'
import { createDefaultInitialGraphSettings } from '~/config/initial'
import type { MarkerConfig } from '~/config/marker'
import { Markers } from '~/config/marker'
import type { Modifiers } from '~/config/modifiers'
import type { PositionInitializer } from '~/config/position'
import { PositionInitializers } from '~/config/position'
import type { SimulationConfig } from '~/config/simulation'
import type { ZoomConfig } from '~/config/zoom'
import type { NodeTypeToken } from '~/model/graph'
import type { GraphLink } from '~/model/link'
import type { GraphNode } from '~/model/node'

export interface GraphConfig<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>
> {
  /**
   * Set to true to enable automatic resizing.
   * Warning: Do call shutdown(), once the controller is no longer required.
   */
  readonly autoResize: boolean
  /**
   * Callback configuration.
   */
  readonly callbacks: Callbacks<T, Node>
  /**
   * Initial settings of a controller.
   */
  readonly initial: InitialGraphSettings<T, Node, Link>
  /**
   * Marker configuration.
   */
  readonly marker: MarkerConfig
  /**
   * Low-level callbacks for modifying the underlying d3-selection.
   */
  readonly modifiers: Modifiers<T, Node, Link>
  /**
   * Define the radius of a node for the simulation and visualization.
   * Can be a static number or a function receiving a node as its parameter.
   */
  readonly nodeRadius: number | ((node: Node) => number)
  /**
   * Initializes a node's position in context of a graph's width and height.
   */
  readonly positionInitializer: PositionInitializer<T, Node>
  /**
   * Simulation configuration.
   */
  readonly simulation: SimulationConfig<T, Node, Link>
  /**
   * Zoom configuration.
   */
  readonly zoom: ZoomConfig
}

function defaultGraphConfig<
  T extends NodeTypeToken,
  Node extends GraphNode<T>,
  Link extends GraphLink<T, Node>
>(): GraphConfig<T, Node, Link> {
  return {
    autoResize: false,
    callbacks: {},
    initial: createDefaultInitialGraphSettings(),
    nodeRadius: 16,
    marker: Markers.Arrow(4),
    modifiers: {},
    positionInitializer: PositionInitializers.Centered,
    simulation: {
      alphas: createDefaultAlphaConfig(),
      forces: createDefaultForceConfig(),
    },
    zoom: {
      initial: 1,
      min: 0.1,
      max: 2,
    },
  }
}

/**
 * Utility type for deeply partial objects.
 */
export type DeepPartial<T> = {
  readonly [P in keyof T]?: DeepPartial<T[P]>
}

/**
 * Define the configuration of a controller.
 * Will be merged with the default configuration.
 * @param config - The partial configuration.
 * @returns The merged configuration.
 */
export function defineGraphConfig<
  T extends NodeTypeToken = NodeTypeToken,
  Node extends GraphNode<T> = GraphNode<T>,
  Link extends GraphLink<T, Node> = GraphLink<T, Node>
>(
  config: DeepPartial<GraphConfig<T, Node, Link>> = {}
): GraphConfig<T, Node, Link> {
  return merge.withOptions(
    { mergeArrays: false },
    defaultGraphConfig<T, Node, Link>(),
    config
  )
}
