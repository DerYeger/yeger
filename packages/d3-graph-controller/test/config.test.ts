import { describe, expect, it } from 'vitest'

import { defineGraphConfig } from '~/config/config'

describe('Config', () => {
  describe('can be defined', () => {
    it('using default values', () => {
      const config = defineGraphConfig()
      expect(config).toMatchSnapshot()
    })

    it('with deep merging', () => {
      const defaultConfig = defineGraphConfig()
      const customConfig = defineGraphConfig({
        simulation: {
          forces: {
            collision: {
              radiusMultiplier: 42,
            },
          },
        },
      })
      const customCollisionForce = customConfig.simulation.forces.collision
      expect(customCollisionForce).not.toBe(false)
      // @ts-expect-error It has been asserted that the force is not false
      expect(customCollisionForce.radiusMultiplier).toEqual(42)
      expect(customConfig).not.toEqual(defaultConfig)

      const customMerge = {
        ...defaultConfig,
        simulation: {
          ...defaultConfig.simulation,
          forces: {
            ...defaultConfig.simulation.forces,
            collision: {
              ...defaultConfig.simulation.forces.collision,
              radiusMultiplier: 42,
            },
          },
        },
      }
      expect(customMerge.simulation.forces).toStrictEqual(
        customConfig.simulation.forces,
      )
    })
  })
})
