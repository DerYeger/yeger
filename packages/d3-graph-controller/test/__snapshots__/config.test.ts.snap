// Vitest Snapshot v1

exports[`Config > can be defined > using default values 1`] = `
{
  "autoResize": false,
  "callbacks": {},
  "initial": {
    "includeUnlinked": true,
    "linkFilter": [Function],
    "nodeTypeFilter": undefined,
    "showLinkLabels": true,
    "showNodeLabels": true,
  },
  "marker": {
    "padding": [Function],
    "path": [
      [
        0,
        0,
      ],
      [
        0,
        4,
      ],
      [
        4,
        2,
      ],
    ],
    "ref": [
      2,
      2,
    ],
    "size": 4,
    "viewBox": "0,0,4,4",
  },
  "modifiers": {},
  "nodeRadius": 16,
  "positionInitializer": [Function],
  "simulation": {
    "alphas": {
      "drag": {
        "end": 0,
        "start": 0.1,
      },
      "filter": {
        "link": 1,
        "type": 0.1,
        "unlinked": {
          "exclude": 0.1,
          "include": 0.1,
        },
      },
      "focus": {
        "acquire": [Function],
        "release": [Function],
      },
      "initialize": 1,
      "labels": {
        "links": {
          "hide": 0,
          "show": 0,
        },
        "nodes": {
          "hide": 0,
          "show": 0,
        },
      },
      "resize": 0.5,
    },
    "forces": {
      "centering": {
        "enabled": true,
        "strength": 0.1,
      },
      "charge": {
        "enabled": true,
        "strength": -1,
      },
      "collision": {
        "enabled": true,
        "radiusMultiplier": 2,
        "strength": 1,
      },
      "link": {
        "enabled": true,
        "length": 128,
        "strength": 1,
      },
    },
  },
  "zoom": {
    "initial": 1,
    "max": 2,
    "min": 0.1,
  },
}
`;
